drivers/fsdiff.go

```diff
@@ -44,6 +44,11 @@ func NewNaiveDiffDriver(driver ProtoDriver, updater LayerIDMapUpdater) Driver {
 func (gdw *NaiveDiffDriver) Diff(id string, idMappings *idtools.IDMappings, parent string, parentMappings *idtools.IDMappings, mountLabel string) (arch io.ReadCloser, err error) {
 	startTime := time.Now()
 	driver := gdw.ProtoDriver
+	// since we are going to remove layers
+	// lock so other parallel instances can
+	// wait while parent layer can be restored.
+	driver.FileLock()
+	defer driver.FileUnlock()
```

Comment:
Is it something we could fix directly in the overlay driver?

And it seems to be locked only here. Is this enough to protect from race conditions?

Reply:
Sure we could also lock here https://github.com/containers/storage/blob/main/drivers/overlay/overlay.go#L1775 my first intention was to make lock granular so we don't end up blocking for operations where lock is not needed.

During my experiments i saw race with fsdiff only so just to make things granular i decided to have it restricted only to fsdiff but i agree we could move this to overlay as well to be extra safe.

Reply:
does the issue happen if there is a fsdiff and at the same time the container/image is being removed?

Reply:
yes this happens whenever flow goes into https://github.com/containers/storage/blob/main/drivers/overlay/overlay.go#L1777 from more than one process and `parent` is same across multiple processes.

So typically one process might be trying to read things here https://github.com/containers/storage/blob/main/drivers/fsdiff.go#L92 while another process removes parent here https://github.com/containers/storage/blob/main/drivers/fsdiff.go#L90

Reply:
why doesn't the locking we have in place work here?

In overlay.go Put() has already a ref counter and it should not do any unmount when there is another user

Reply:
I believe that is because `Put` has a mutex lock which only works for goroutines and thread in same process. However we do have a `fs` lock in `unmount` but at one place we are removing directory outside `fs` lock or `unmount` i.e https://github.com/containers/storage/blob/main/drivers/overlay/overlay.go#L1564 which is causing issues.

Reply:
I think this is there in the first place because most of driver code is fetched from **moby** but they dont have inter-process race situations since everything works from a single daemon process.